Problem with deriving schema for type tree with two traits as union and interface

Hi, I have encountered an error with the Schema generation and succeeded in reproducing it. The Schema is defined as follows:

```scala
  case class C(c: A)

  @GQLInterface
  sealed trait A {
    def a: Int
    def z: C
  }

  sealed trait B

  case class A1(a: Int, z: C) extends A
  case class B1(a: Int, b: Int, z: C) extends A with B
  case class B2(a: Int, b: Int, z: C) extends A with B

  case class Mutation(
    a: Int => A1,
    b: Int => B,
)
```

It results in the following Schema:

```
interface A {
  a: Int!
  z: C!
}

type A1 {
  a: Int!
  z: C!
}

union B = B1 | B2
type B1 implements A {
  a: Int!
  b: Int!
  z: C!
}
​
type B2 implements A {
  a: Int!
  b: Int!
  z: C!
}
​
type C {
  c: A!
}
```

The important part here is that `A1` doesn't implement `A` interface in the generated schema - despite specifying that A1 implements A in Scala code. While reproducing this error I've experimented a bit and I may've got some idea about this issue that may help you resolve it. If I introduce a mutation that uses `A` type before other mutations in `Mutation` case class, the issue is fixed and A1 derives from A. And by `before` I mean introducing it as a first field in Mutation object - order is important here. Moreover, explicitly writing something like `implicit val xyz = gen[A]` or `implicit val xyz = gen[A1]` doesn't help at all. From my understanding, the reason for all this is that the type `A` is derived after type `A1` (after deriving nested `C`), and all that while deriving the `Mutation` type. What I ended up doing in my case is just having a `dummy` mutation as first field in `Mutation` object that uses `A` type:
```
    @GQLDescription("Dummy value introduced to fix Caliban type derivation error. Will throw an exception when called.")
    dummy: Int => A,
```
It's definitely not perfect and I would be really thankful if you could help me :)

I'm using scala 2.13 and I've tried to call sbt +publishLocal. However, I ended up having some errors thrown end everything ended with:
```
[error] (core / Compile / compileIncremental) java.lang.AssertionError: assertion failed: ClassBType.info not yet assigned: Ljava/lang/Object;
[error] (clientJS / Compile / compileIncremental) java.lang.AssertionError: assertion failed: ClassBType.info not yet assigned: Ljava/lang/Object;
[error] (clientJVM / Compile / compileIncremental) java.lang.AssertionError: assertion failed: ClassBType.info not yet assigned: Ljava/lang/Object;
```
I rarely use sbt and I may've got something wrong. Do you know what could be causing it? You could also just upload the built jar for me if it would be a faster way for you.

--- a/core/src/main/scala/caliban/Rendering.scala
+++ b/core/src/main/scala/caliban/Rendering.scala
@@ -52,7 +52,7 @@ object Rendering {
     t.interfaces()
       .fold("")(_.flatMap(_.name) match {
         case Nil  => ""
-        case list => s" implements ${list.mkString("& ")}"
+        case list => s" implements ${list.mkString(" & ")}"
       })
 
   private def renderKind(kind: __TypeKind): String =

--- a/core/src/main/scala/caliban/schema/Types.scala
+++ b/core/src/main/scala/caliban/schema/Types.scala
@@ -101,7 +101,14 @@ object Types {
         t.ofType.fold(existingTypes)(collectTypes(_, existingTypes))
       case _ =>
         val list1 =
-          t.name.fold(existingTypes)(_ => if (existingTypes.exists(same(t, _))) existingTypes else t :: existingTypes)
+          t.name.fold(existingTypes)(_ =>
+            if (existingTypes.exists(same(t, _))) {
+              existingTypes.map {
+                case ex if ex.name == t.name && t.interfaces().nonEmpty => t
+                case other                                              => other
+              }
+            } else t :: existingTypes
+          )
         val embeddedTypes =
           t.fields(__DeprecatedArgs(Some(true))).getOrElse(Nil).flatMap(f => f.`type` :: f.args.map(_.`type`)) ++
             t.inputFields.getOrElse(Nil).map(_.`type`)

--- a/core/src/main/scala/caliban/schema/Types.scala
+++ b/core/src/main/scala/caliban/schema/Types.scala
@@ -104,8 +104,18 @@ object Types {
           t.name.fold(existingTypes)(_ =>
             if (existingTypes.exists(same(t, _))) {
               existingTypes.map {
-                case ex if ex.name == t.name && t.interfaces().nonEmpty => t
-                case other                                              => other
+                case ex if same(ex, t) =>
+                  ex.copy(interfaces =
+                    () =>
+                      (ex.interfaces(), t.interfaces()) match {
+                        case (None, None)             => None
+                        case (Some(interfaces), None) => Some(interfaces)
+                        case (None, Some(interfaces)) => Some(interfaces)
+                        case (Some(left), Some(right)) =>
+                          Some(left ++ right.filterNot(t => left.exists(_.name == t.name)))
+                      }
+                  )
+                case other => other
               }
             } else t :: existingTypes
           )

